File: src/com/android/launcher3/views/Snackbar.java
/*
 * Copyright (C) 2018 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.android.launcher3.views;

import static android.view.accessibility.AccessibilityManager.FLAG_CONTENT_CONTROLS;
import static android.view.accessibility.AccessibilityManager.FLAG_CONTENT_TEXT;

import android.content.Context;
import android.content.res.Resources;
import android.graphics.Rect;
import android.util.AttributeSet;
import android.util.TypedValue;
import android.view.Gravity;
import android.view.MotionEvent;
import android.widget.TextView;

import androidx.annotation.Nullable;

import com.android.launcher3.AbstractFloatingView;
import com.android.launcher3.DeviceProfile;
import com.android.launcher3.R;
import com.android.launcher3.anim.Interpolators;
import com.android.launcher3.compat.AccessibilityManagerCompat;
import com.android.launcher3.dragndrop.DragLayer;

/**
 * A toast-like UI at the bottom of the screen with a label, button action, and dismiss action.
 */
public class Snackbar extends AbstractFloatingView {

    private static final long SHOW_DURATION_MS = 180;
    private static final long HIDE_DURATION_MS = 180;
    private static final int TIMEOUT_DURATION_MS = 4000;

    private final ActivityContext mActivity;
    private Runnable mOnDismissed;

    public Snackbar(Context context, AttributeSet attrs) {
        this(context, attrs, 0);
    }

    public Snackbar(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        mActivity = ActivityContext.lookupContext(context);
        inflate(context, R.layout.snackbar, this);
    }

    /** Show a snackbar with just a label. */
    public static <T extends Context & ActivityContext> void show(T activity, int labelStringRedId,
            Runnable onDismissed) {
        show(activity, labelStringRedId, NO_ID, onDismissed, null);
    }

    /** Show a snackbar with a label and action. */
    public static <T extends Context & ActivityContext> void show(T activity, int labelStringResId,
            int actionStringResId, Runnable onDismissed, @Nullable Runnable onActionClicked) {
        closeOpenViews(activity, true, TYPE_SNACKBAR);
        Snackbar snackbar = new Snackbar(activity, null);
        // Set some properties here since inflated xml only contains the children.
        snackbar.setOrientation(HORIZONTAL);
        snackbar.setGravity(Gravity.CENTER_VERTICAL);
        Resources res = activity.getResources();
        snackbar.setElevation(res.getDimension(R.dimen.snackbar_elevation));
        int padding = res.getDimensionPixelSize(R.dimen.snackbar_padding);
        snackbar.setPadding(padding, padding, padding, padding);
        snackbar.setBackgroundResource(R.drawable.round_rect_primary);

        snackbar.mIsOpen = true;
        BaseDragLayer dragLayer = activity.getDragLayer();
        dragLayer.addView(snackbar);

        DragLayer.LayoutParams params = (DragLayer.LayoutParams) snackbar.getLayoutParams();
        params.gravity = Gravity.CENTER_HORIZONTAL | Gravity.BOTTOM;
        params.height = res.getDimensionPixelSize(R.dimen.snackbar_height);
        int maxMarginLeftRight = res.getDimensionPixelSize(R.dimen.snackbar_max_margin_left_right);
        int minMarginLeftRight = res.getDimensionPixelSize(R.dimen.snackbar_min_margin_left_right);
        int marginBottom = res.getDimensionPixelSize(R.dimen.snackbar_margin_bottom);
        int absoluteMaxWidth = res.getDimensionPixelSize(R.dimen.snackbar_max_width);
        Rect insets = activity.getDeviceProfile().getInsets();
        int maxWidth = Math.min(
                dragLayer.getWidth() - minMarginLeftRight * 2 - insets.left - insets.right,
                absoluteMaxWidth);
        int minWidth = Math.min(
                dragLayer.getWidth() - maxMarginLeftRight * 2 - insets.left - insets.right,
                absoluteMaxWidth);
        params.width = minWidth;
        DeviceProfile deviceProfile = activity.getDeviceProfile();
        params.setMargins(0, 0, 0, marginBottom
                + (deviceProfile.isTaskbarPresent
                ? deviceProfile.taskbarSize + deviceProfile.getTaskbarOffsetY()
                : insets.bottom));

        TextView labelView = snackbar.findViewById(R.id.label);
        String labelText = res.getString(labelStringResId);
        labelView.setText(labelText);

        TextView actionView = snackbar.findViewById(R.id.action);
        float actionWidth;
        if (actionStringResId != NO_ID) {
            String actionText = res.getString(actionStringResId);
            actionWidth = actionView.getPaint().measureText(actionText)
                    + actionView.getPaddingRight() + actionView.getPaddingLeft();
            actionView.setText(actionText);
            actionView.setOnClickListener(v -> {
                if (onActionClicked != null) {
                    onActionClicked.run();
                }
                snackbar.mOnDismissed = null;
                snackbar.close(true);
            });
        } else {
            actionWidth = 0;
            actionView.setVisibility(GONE);
        }

        int totalContentWidth = (int) (labelView.getPaint().measureText(labelText) + actionWidth)
                + labelView.getPaddingRight() + labelView.getPaddingLeft()
                + padding * 2;
        if (totalContentWidth > params.width) {
            // The text doesn't fit in our standard width so update width to accommodate.
            if (totalContentWidth <= maxWidth) {
                params.width = totalContentWidth;
            } else {
                // One line will be cut off, fallback to 2 lines and smaller font. (This should only
                // happen in some languages if system display and font size are set to largest.)
                int textHeight = res.getDimensionPixelSize(R.dimen.snackbar_content_height);
                float textSizePx = res.getDimension(R.dimen.snackbar_min_text_size);
                labelView.setLines(2);
                labelView.getLayoutParams().height = textHeight * 2;
                actionView.getLayoutParams().height = textHeight * 2;
                labelView.setTextSize(TypedValue.COMPLEX_UNIT_PX, textSizePx);
                actionView.setTextSize(TypedValue.COMPLEX_UNIT_PX, textSizePx);
                params.height += textHeight;
                params.width = maxWidth;
            }
        }

        snackbar.mOnDismissed = onDismissed;
        snackbar.setAlpha(0);
        snackbar.setScaleX(0.8f);
        snackbar.setScaleY(0.8f);
        snackbar.animate()
                .alpha(1f)
                .withLayer()
                .scaleX(1)
                .scaleY(1)
                .setDuration(SHOW_DURATION_MS)
                .setInterpolator(Interpolators.ACCEL_DEACCEL)
                .start();
        int timeout = AccessibilityManagerCompat.getRecommendedTimeoutMillis(activity,
                TIMEOUT_DURATION_MS, FLAG_CONTENT_TEXT | FLAG_CONTENT_CONTROLS);
        snackbar.postDelayed(() -> snackbar.close(true), timeout);
    }

    @Override
    protected void handleClose(boolean animate) {
        if (mIsOpen) {
            if (animate) {
                animate().alpha(0f)
                        .withLayer()
                        .setStartDelay(0)
                        .setDuration(HIDE_DURATION_MS)
                        .setInterpolator(Interpolators.ACCEL)
                        .withEndAction(this::onClosed)
                        .start();
            } else {
                animate().cancel();
                onClosed();
            }
            mIsOpen = false;
        }
    }

    private void onClosed() {
        mActivity.getDragLayer().removeView(this);
        if (mOnDismissed != null) {
            mOnDismissed.run();
        }
    }

    @Override
    protected boolean isOfType(int type) {
        return (type & TYPE_SNACKBAR) != 0;
    }

    @Override
    public boolean onControllerInterceptTouchEvent(MotionEvent ev) {
        if (ev.getAction() == MotionEvent.ACTION_DOWN) {
            BaseDragLayer dl = mActivity.getDragLayer();
            if (!dl.isEventOverView(this, ev)) {
                close(true);
            }
        }
        return false;
    }
}

File: lawnchair/src/ch/deletescape/lawnchair/views/Snackbar.kt
/*
 *     This file is part of Lawnchair Launcher.
 *
 *     Lawnchair Launcher is free software: you can redistribute it and/or modify
 *     it under the terms of the GNU General Public License as published by
 *     the Free Software Foundation, either version 3 of the License, or
 *     (at your option) any later version.
 *
 *     Lawnchair Launcher is distributed in the hope that it will be useful,
 *     but WITHOUT ANY WARRANTY; without even the implied warranty of
 *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *     GNU General Public License for more details.
 *
 *     You should have received a copy of the GNU General Public License
 *     along with Lawnchair Launcher.  If not, see <https://www.gnu.org/licenses/>.
 */

package ch.deletescape.lawnchair.views

import android.content.Context
import android.util.AttributeSet
import android.view.MotionEvent
import android.view.View
import android.widget.LinearLayout
import android.widget.TextView
import ch.deletescape.lawnchair.colors.ColorEngine
import com.android.launcher3.AbstractFloatingView
import com.android.launcher3.Launcher
import com.android.launcher3.R
import com.android.launcher3.anim.Interpolators
import com.android.launcher3.views.BaseDragLayer

class Snackbar(context: Context, attributeSet: AttributeSet?) : AbstractFloatingView(context, attributeSet) {

    private val launcher = Launcher.getLauncher(context)
    var onDismissed: Runnable? = null

    init {
        LinearLayout.inflate(context, R.layout.snackbar, this)
    }

    public override fun handleClose(z: Boolean) {
        if (mIsOpen) {
            if (z) {
                animate()
                        .alpha(0f)
                        .withLayer()
                        .setStartDelay(0)
                        .setDuration(180)
                        .setInterpolator(Interpolators.ACCEL)
                        .withEndAction(::onClosed).start()
            } else {
                animate().cancel()
                onClosed()
            }
            this.mIsOpen = false
        }
    }

    public override fun isOfType(i: Int): Boolean {
        return i and 128 != 0
    }

    override fun logActionCommand(i: Int) {}

    private fun onClosed() {
        launcher.dragLayer.removeView(this)
        onDismissed?.run()
    }

    override fun onControllerInterceptTouchEvent(motionEvent: MotionEvent): Boolean {
        if (motionEvent.action == 0 && !launcher.dragLayer.isEventOverView(this, motionEvent)) {
            close(true)
        }
        return false
    }

    companion object {

        @JvmStatic
        fun show(launcher: Launcher, i: Int, i2: Int, runnable: Runnable?, runnable2: Runnable?) {
            AbstractFloatingView.closeOpenViews(launcher, true, AbstractFloatingView.TYPE_QUICKSTEP_PREVIEW)
            val snackbar = Snackbar(launcher, null)
            snackbar.orientation = LinearLayout.HORIZONTAL
            snackbar.gravity = 16
            val resources = launcher.resources
            snackbar.elevation = resources.getDimension(R.dimen.snackbar_elevation)
            var dimensionPixelSize = resources.getDimensionPixelSize(R.dimen.snackbar_padding)
            snackbar.setPadding(dimensionPixelSize, dimensionPixelSize, dimensionPixelSize, dimensionPixelSize)
            snackbar.setBackgroundResource(R.drawable.round_rect_primary)
            snackbar.mIsOpen = true
            val dragLayer = launcher.dragLayer
            dragLayer.addView(snackbar)
            val layoutParams = snackbar.layoutParams as BaseDragLayer.LayoutParams
            layoutParams.gravity = 81
            layoutParams.height = resources.getDimensionPixelSize(R.dimen.snackbar_height)
            val dimensionPixelSize2 = resources.getDimensionPixelSize(R.dimen.snackbar_max_margin_left_right)
            var dimensionPixelSize3 = resources.getDimensionPixelSize(R.dimen.snackbar_min_margin_left_right)
            val dimensionPixelSize4 = resources.getDimensionPixelSize(R.dimen.snackbar_margin_bottom)
            val rect = launcher.deviceProfile.insets
            val width = dragLayer.width - dimensionPixelSize3 * 2 - rect.left - rect.right
            layoutParams.width = dragLayer.width - dimensionPixelSize2 * 2 - rect.left - rect.right
            layoutParams.setMargins(0, 0, 0, dimensionPixelSize4 + rect.bottom)
            val textView = snackbar.findViewById<View>(R.id.label) as TextView
            val textView2 = snackbar.findViewById<View>(R.id.action) as TextView
            val string = resources.getString(i)
            val string2 = resources.getString(i2)
            dimensionPixelSize *= 2
            dimensionPixelSize += textView2.paddingLeft + (textView2.paddingRight + (textView.paddingLeft + (textView.paddingRight + (textView2.paint.measureText(string2) + textView.paint.measureText(string)).toInt())))
            if (dimensionPixelSize > layoutParams.width) {
                if (dimensionPixelSize <= width) {
                    layoutParams.width = dimensionPixelSize
                } else {
                    dimensionPixelSize = resources.getDimensionPixelSize(R.dimen.snackbar_content_height)
                    val dimension = resources.getDimension(R.dimen.snackbar_min_text_size)
                    textView.setLines(2)
                    dimensionPixelSize3 = dimensionPixelSize * 2
                    textView.layoutParams.height = dimensionPixelSize3
                    textView2.layoutParams.height = dimensionPixelSize3
                    textView.setTextSize(0, dimension)
                    textView2.setTextSize(0, dimension)
                    layoutParams.height += dimensionPixelSize
                    layoutParams.width = width
                }
            }
            textView.text = string
            textView2.text = string2
            textView2.setTextColor(ColorEngine.getInstance(launcher).accent)
            textView2.setOnClickListener {
                runnable2?.run()
                snackbar.onDismissed = null
                snackbar.close(true)
            }
            snackbar.onDismissed = runnable
            snackbar.alpha = 0f
            snackbar.scaleX = 0.8f
            snackbar.scaleY = 0.8f
            snackbar.animate().alpha(1f).withLayer().scaleX(1f).scaleY(1f).setDuration(180).setInterpolator(Interpolators.ACCEL_DEACCEL).start()
            snackbar.postDelayed({ snackbar.close(true) }, 4000)
        }
    }
}